### ISSUE TITLE
The key and value types can't be covariant

### ISSUE THREAD

**Issue by steffahn** [ibraheemdev/papaya#41]
The key and value types can't be covariant, for API soundness

e.g. here’s a bad use-case
```rust
use papaya::HashMap;

// abuse variance of V
fn extended_lifetime_v(x: &str) -> &'static str {
    let m = HashMap::new();
    m.pin().insert((), x);
    let extended: &str = m.pin().get(&()).unwrap();
    extended
}

// abuse variance of K
fn extended_lifetime_k(x: &str) -> &'static str {
    let m = HashMap::new();
    m.pin().insert(x, ());
    let extended: &str = m.pin().get_key_value(x).unwrap().0;
    extended
}

fn main() {
    let s = "Hello dear friend of use-after-free!".to_owned();
    let r1 = extended_lifetime_k(&s); // either
    let r2 = extended_lifetime_v(&s); // works
    drop(s);
    println!("{r1}");
    println!("{r2}");
}
```

**Comment by ibraheemdev** [comment]
Hmm, thanks for catching this. Would `T: 'static` be an appropriate fix?

**Comment by steffahn** [comment]
No, the way to make them invariant would be to replace the `PhantomData<(K, V)>` in `raw::HashMap` with something like `PhantomData<Cell<(K, V)>>`.

`T: 'static` bounds actually (surprisingly) don't help with variance, as you can still have subtyping relationships such as between `Box<dyn for<'a> Fn(&'a str)>` and `Box<dyn Fn(&'static str)>`.

**Comment by steffahn** [comment]
While I'm at it, I'd also like to point out the `Sync` implementation as problematic. I assume sharing a `HashMap<K, V, …>` can result in the destructor of `K` or `V` values running in a different thread than where they were constructed; in which case `HashMap<K, V, …>: Sync` should not just require `K: Sync, V: Sync`, but actually `K: Send + Sync`, `V: Send + Sync`.

**Comment by ibraheemdev** [comment]
Ah yes, that is true. I believe both of these issues apply to [`flurry`](https://github.com/jonhoo/flurry) well.

**Comment by steffahn** [comment]
Just looking at the API docs, for flurry, it seems like their internal usage of `AtomicPtr` should ensure invariance, and their insert-like methods all require `Send + Sync` for `K` and `V`. I haven't tested it though.

It does make me wonder though; these `Send + Sync` bounds are effectively even somewhat *more* restrictive (you can't really use the map *at all* with non threadsafe types); if there's something about the reclamation via seize that makes simply constraining the `Sync` implementation of the map itself insufficient.

**Comment by ibraheemdev** [comment]
Ah, I see why this is an issue now. The phantom generics `K` and `V` should really be on `raw::Table`, which is what is stored behind `AtomicPtr`, to ensure invariance.

I do think flurry's bounds are overly restrictive, seize doesn't internally spawn any threads, so the `Send` bound would only be necessary if you are able to initiate reclamation from a different thread, i.e. the `HashMap` is `Send` or `Sync`. This is probably a remnant of when flurry used crossbeam-epoch.

**Comment by ibraheemdev** [comment]
Actually I am unsure about this now, but I still think it is correct. If `HashMap` exposed access to the inner `Collector`, it would be possible to cause reclamation on another thread without actually sharing the map. Neither flurry nor papaya do this, and the map owns the collector to which active threads are scoped, so I think this is avoided. Additionally, any guards created before transferring ownership of the collector the map must have been dropped or leaked, so no other thread can have been active once the map is created.

**Comment by steffahn** [comment]
I think so, too. The thread-locals all seem contained to the object, so completely single-threaded use would never involve any other threads; and the only other remaining question would be whether `HashMap: Send` needs `Sync` on `K` or `V`, but with the lifetime on the guards, I don’t see any way in which any shared access to the `K` or `V` could come about in such a case, either. So it seem alright to just to the `K & V: Send+Sync` for `HashMap: Send`

**Comment by ibraheemdev** [comment]
Released the fixes in 0.1.7. Thanks for the report!
